README
NAME
    Date::Simple - a simple date object

SYNOPSIS
        use Date::Simple ('date', 'today');

        # Difference in days between two dates:
        $diff = date('2001-08-27') - date('1977-10-05');

        # Offset $n days from now:
        $date = today() + $n;
        print "$date\n";  # uses ISO 8601 format (YYYY-MM-DD)

        use Date::Simple ();
        my $date  = Date::Simple->new('1972-01-17');
        my $year  = $date->year;
        my $month = $date->month;
        my $day   = $date->day;

        use Date::Simple (':all');
        my $date2 = ymd($year, $month, $day);
        my $date3 = d8('19871218');
        my $today = today();
        my $tomorrow = $today + 1;
        if ($tomorrow->year != $today->year) {
            print "Today is New Year's Eve!\n";
        }

        if ($today > $tomorrow) {
            die "warp in space-time continuum";
        }

        print "Today is ";
        print(('Sun','Mon','Tues','Wednes','Thurs','Fri','Satur')
              [$today->day_of_week]);
        print "day.\n";

        # you can also do this:
        ($date cmp "2001-07-01")
        # and this
        ($date <=> [2001, 7, 1])

INSTALLATION

     If your system has the "make" program or a clone:

         perl Makefile.PL
         make
         make test
         make install

     If you lack "make", copy the "lib/Date" directory to your module
     directory (run "perl -V:sitelib" to find it).

     If "make test" fails, perhaps it means your system can't compile C
     code.  Try:

         make distclean
         perl Makefile.PL noxs
         make
         make test
         make install

     This will use the pure-Perl implementation.

DESCRIPTION
    Dates are complex enough without times and timezones. This module may be
    used to create simple date objects. It handles:

    Validation.
        Reject 1999-02-29 but accept 2000-02-29.

    Interval arithmetic.
        How many days were between two given dates? What date comes N days
        after today?

    Day-of-week calculation.
        What day of the week is a given date?

    It does not deal with hours, minutes, seconds, and time zones.

    A date is uniquely identified by year, month, and day integers within
    valid ranges. This module will not allow the creation of objects for
    invalid dates. Attempting to create an invalid date will return undef.
    Month numbering starts at 1 for January, unlike in C and Java. Years are
    4-digit.

    Gregorian dates up to year 9999 are handled correctly, but we rely on
    Perl's builtin "localtime" function when the current date is requested.
    On some platforms, "localtime" may be vulnerable to rollovers such as
    the Unix "time_t" wraparound of 18 January 2038.

    Overloading is used so you can compare or subtract two dates using
    standard numeric operators such as "==", and the sum of a date object
    and an integer is another date object.

    Date::Simple objects are immutable. After assigning "$date1" to
    "$date2", no change to "$date1" can affect "$date2". This means, for
    example, that there is nothing like a "set_year" operation, and
    "$date++" assigns a new object to "$date".

    This module contains various undocumented functions. They may not be
    available on all platforms and are likely to change or disappear in
    future releases. Please let the author know if you think any of them
    should be public.

CONSTRUCTORS
    Several functions take a string or numeric representation and generate a
    corresponding date object. The most general is "new", whose argument
    list may be empty (returning the current date), a string in format
    YYYY-MM-DD or YYYYMMDD, a list or arrayref of year, month, and day
    number, or an existing date object.

    Date::Simple->new ([ARG, ...])
    date ([ARG, ...])
            my $date = Date::Simple->new('1972-01-17');

        The "new" method will return a date object if the values passed in
        specify a valid date. (See above.) If an invalid date is passed, the
        method returns undef. If the argument is invalid in form as opposed
        to numeric range, "new" dies.

        The "date" function provides the same functionality but must be
        imported or qualified as "Date::Simple::date". (To import all public
        functions, do "use Date::Simple (':all');".) This function returns
        undef on all invalid input, rather than dying in some cases like
        "new".

    today()
        Returns the current date according to "localtime".

        Caution: To get tomorrow's date (or any fixed offset from today), do
        not use "today + 1". Perl parses this as "today(+1)". You need to
        put empty parentheses after the function: "today() + 1".

    ymd (YEAR, MONTH, DAY)
        Returns a date object with the given year, month, and day numbers.
        If the arguments do not specify a valid date, undef is returned.

        Example:

            use Date::Simple ('ymd');
            $pbd = ymd(1987, 12, 18);

    d8 (STRING)
        Parses STRING as "YYYYMMDD" and returns the corresponding date
        object, or undef if STRING has the wrong format or specifies an
        invalid date.

        Example:

            use Date::Simple ('d8');
            $doi = d8('17760704');

        Mnemonic: The string matches "/\d{8}/". Also, "d8" spells "date", if
        8 is expanded phonetically.

INSTANCE METHODS
    DATE->next
            my $tomorrow = $today->next;

        Returns an object representing tomorrow.

    DATE->prev
            my $yesterday = $today->prev;

        Returns an object representing yesterday.

    DATE->year
            my $year  = $date->year;

        Return the year of DATE as an integer.

    DATE->month
            my $month = $date->month;

        Return the month of DATE as an integer from 1 to 12.

    DATE->day
            my $day   = $date->day;

        Return the DATE's day of the month as an integer from 1 to 31.

    DATE->day_of_week
        Return a number representing DATE's day of the week from 0 to 6,
        where 0 means Sunday.

    DATE->as_ymd
            my ($year, $month, $day) = $date->as_ymd;

        Returns a list of three numbers: year, month, and day.

    DATE->as_d8
        Returns the "d8" representation (see "d8"), like
        "$date->format("%Y%m%d")".

    DATE->format (STRING)
    DATE->strftime (STRING)
        These functions are equivalent. Return a string representing the
        date, in the format specified. If you don't pass a parameter, an ISO
        8601 formatted date is returned.

            my $change_date = $date->format("%d %b %y");
            my $iso_date1 = $date->format("%Y-%m-%d");
            my $iso_date2 = $date->format;

        The formatting parameter is similar to one you would pass to
        strftime(3). This is because we actually do pass it to strftime to
        format the date. This may result in differing behavior across
        platforms and locales and may not even work everywhere.

OPERATORS
    Some operators can be used with Date::Simple instances. If one side of
    an expression is a date object, and the operator expects two date
    objects, the other side is interpreted as "date(ARG)", so an array
    reference or ISO 8601 string will work.

    DATE + NUMBER
    DATE - NUMBER
        You can construct a new date offset by a number of days using the
        "+" and "-" operators.

    DATE1 - DATE2
        You can subtract two dates to find the number of days between them.

    DATE1 == DATE2
    DATE1 < DATE2
    DATE1 <=> DATE2
    DATE1 cmp DATE2
    etc.
        You can compare two dates using the arithmetic or string comparison
        operators. Equality tests ("==" and "eq") return false when one of
        the expressions can not be converted to a date. Other comparison
        tests die in such cases. This is intentional, because in a sense,
        all non-dates are not "equal" to all dates, but in no sense are they
        "greater" or "less" than dates.

    DATE += NUMBER
    DATE -= NUMBER
        You can increment or decrement a date by a number of days using the
        += and -= operators. This actually generates a new date object and
        is equivalent to "$date = $date + $number".

    "$date"
        You can interpolate a date instance directly into a string, in the
        format specified by ISO 8601 (eg: 2000-01-17).

UTILITIES
    leap_year (YEAR)
        Returns true if YEAR is a leap year.

    days_in_month (YEAR, MONTH)
        Returns the number of days in MONTH, YEAR.

AUTHOR
        Marty Pauley <marty@kasei.com>
        John Tobey <jtobey@john-edwin-tobey.org>

COPYRIGHT
          Copyright (C) 2001  Kasei
          Copyright (C) 2001-2004 John Tobey.

          This program is free software; you can redistribute it and/or
          modify it under the terms of either:

          a) the GNU General Public License;
             either version 2 of the License, or (at your option) any later
             version.  You should have received a copy of the GNU General
             Public License along with this program; see the file COPYING.
             If not, write to the Free Software Foundation, Inc., 59
             Temple Place, Suite 330, Boston, MA 02111-1307 USA

          b) the Perl Artistic License.

          This program is distributed in the hope that it will be useful,
          but WITHOUT ANY WARRANTY; without even the implied warranty of
          MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.



ChangeLog
2008-01-10  Igor Sutton <igor.sutton@yahoo.co.uk>

	* Copy default_format from old Date::Simple object when adding or
	subtracting (bug #31037).
	
2008-01-09  Igor Sutton <igor.sutton@yahoo.co.uk>

	* Applied patch proposed for 64bit support (bug #28549).

2004-12-18  Yves Orton <demerphq@hotmail.com>

        * YO --  Version number kick to 3.02, doc fix. Patch merge.

2004-11-27  John Tobey  <jtobey@john-edwin-tobey.org>

	* Simple.xs (is_leap_year): Fixed math for dates before
	1900-03-01.  Thanks to Anil Natha for finding the bug.

	* t/date.t: Added day_of_week tests.


2004-04-02  Yves Orton <demerphq@hotmail.com>

	* lib/Date/Simple.pm ($VERSION): Version 3.
	  Added formatting support and as_iso() XS method.
	  Switched to using Test::More instead of Test.
	  Prereqs expanded to include Test::More and Scalar::Util
	  Both are standard distro in later perls.

	* Makefile.pl: Dont mess with filenames xs is true.

2002-12-06  John Tobey  <jtobey@john-edwin-tobey.org>

	* lib/Date/Simple.pm ($VERSION): Version 2.04.

	* lib/Date/Simple.pm (use overload): Overload more operators: ==
	!= eq ne
	(_inval): Stylistic changes.
	(_new): New sub, built from the old 'new'.  Accept YYYYMMDD
	strings without hyphens.  Return undef rather than dying on
	unrecognized format.  ('new' still dies for compatibility.)
	(date): Use '_new'.  Avoid use of 'eval' and $@.
	(new): Use '_new'.
	(pod): Update docs.

	* lib/Date/Simple/NoXS.pm (_eq): New sub.
	(_ne): New sub.

	* Simple.xs (is_object): New function.
	(new_for_cmp): New function.
	(_add): Use is_object().
	(_subtract): Use is_object().
	(_compare): Use is_object() and new_for_cmp().
	(_eq): New sub.
	(_ne): New sub.

	* t/date.t: Test new equality operators.

	* README: Regenerated from lib/Date/Simple.pm.

2002-03-11  John Tobey  <jtobey@john-edwin-tobey.org>

	* lib/Date/Simple.pm ($VERSION): Version 2.03.

	* t/date.t: Remove a locale dependency.  Thanks to ASSAD Arnaud
	for the fix.

2001-10-02  John Tobey  <jtobey@john-edwin-tobey.org>

	* lib/Date/Simple.pm ($VERSION): Version 2.02.

	* Simple.xs (days_in_month): avoid crash due to reversed args.

2001-08-27  John Tobey  <jtobey@john-edwin-tobey.org>

	* lib/Date/Simple.pm: Version 2.01.

	* t/date.t: Test today() fix.

	* lib/Date/Simple.pm (today): Replace use of deleted function.

	* Version 2.00.

	* Added functional interface (date, d8, ymd).

	* Load POSIX module only when needed.

	* Crunch numbers in C if compiler found.  (Install with 'perl
	Makefile.PL noxs' to select the pure Perl implementation.)

	* Can calculate the day of the week.

	* Reduced dependence on time_t, most functions work for dates up
	to 9999.


lib/Date/Simple/NoXS.pm
# Date::Simple::NoXS - used internally by Date::Simple.

use strict;

package Date::Simple;

sub _ymd {
    my ($o,@args)=@_;
    my $c=ref($o)||$o;
    my $days = ymd_to_days(@args);
    return undef unless defined ($days);
    return (bless (\$days, $c));
}

sub _d8 {
    my ($o,$d8) = @_;
    my @ymd = $d8 =~ m/^(\d{4})(\d\d)(\d\d)$/ or return undef;
    return $o->_ymd(@ymd);
}

# Precise integer arithmetic functions unfortunately missing from
# Perl's core:

sub _divmod {
    my ($quot, $int);

    $quot = $_[0] / $_[1];
    $int = int($quot);
    $int -= 1 if $int > $quot;
    $_[0] %= $_[1];
    return $int;
};

sub _div {
    my ($quot, $int);

    $quot = $_[0] / $_[1];
    $int = int($quot);
    return $int - 1 if $int > $quot;
    return $int;
};

sub leap_year {
    my $y = shift;
    return (($y%4==0) and ($y%400==0 or $y%100!=0)) || 0;
}

my @days_in_month = (
 [0,31,28,31,30,31,30,31,31,30,31,30,31],
 [0,31,29,31,30,31,30,31,31,30,31,30,31],
);

sub days_in_month ($$) {
    my ($y,$m) = @_;
    return $days_in_month[leap_year($y)][$m];
}

sub validate ($$$) {
    my ($y, $m, $d)= @_;
    # any +ve integral year is valid
    return 0 if $y != abs int $y;
    return 0 unless 1 <= $m and $m <= 12;
    return 0 unless 1 <= $d and $d <= $days_in_month[leap_year($y)][$m];
    return 1;
}

# Given a year, month, and day, return the canonical day number.
# That is the number of days since 1 January 1970, negative if earlier.
sub ymd_to_days {
    my ($Y, $M, $D) = @_;
    my ($days, $x);

    if ($M < 1 || $M > 12 || $D < 1 ||
	($D > 28 && $D > days_in_month($Y, $M)))
    {
	return undef;
    }

    $days = $D +
	(undef, -1, 30, 58, 89, 119, 150, 180, 211, 242, 272, 303, 333)[$M];
    $days += 365 * ($Y - 1970);
    $x = ($M <= 2 ? $Y-1 : $Y);
    $days += _div (($x - 1968), 4);
    $days -= _div (($x - 1900), 100);
    $days += _div (($x - 1600), 400);
    return $days;
}

sub days_since_1970 { ${$_[0]} }

# Given a canonical day number (days since 1 Jan 1970), return the
# year, month, and day.
sub days_to_ymd {
    my ($days) = @_;
    my ($year, $mnum, $mday, $tmp);

    # Shift frame of reference from 1 Jan 1970 to (the imaginary) 1 Mar 0AD.
    $tmp = $days + 719468;

    # Do the math.
    $year = 400 * _divmod ($tmp, 146097);
    if ($tmp == 146096) {
	# Handle 29 Feb 2000, 2400, ...
	$year += 400;
	$mnum = 2;
	$mday = 29;
    } else {
	$year += 100 * _divmod ($tmp, 36524);
	$year += 4 * _divmod ($tmp, 1461);
	if ($tmp == 1460) {
	    $year += 4;
	    $mnum = 2;
	    $mday = 29;
	} else {
	    $year += _divmod ($tmp, 365);
	    $mnum = _divmod ($tmp, 31);
	    $mday = $tmp + (1, 1, 2, 2, 3, 3, 3, 4, 4, 5, 5, 5)[$mnum];
	    $tmp = (31, 30, 31, 30, 31, 31, 30, 31, 30, 31, 31, 28)[$mnum];
	    if ($mday > $tmp) {
		$mday -= $tmp;
		$mnum += 1;
	    }
	    if ($mnum > 9) {
		$mnum -= 9;
		$year += 1;
	    } else {
		$mnum += 3;
	    }
	}
    }
    return ($year, $mnum, $mday);
}

sub as_ymd { return days_to_ymd (${$_[0]}); }
sub as_d8  { return sprintf ("%04d%02d%02d", &as_ymd); }
sub as_iso { return sprintf ("%04d-%02d-%02d", &as_ymd); }

sub year  { return (&as_ymd) [0]; }
sub month { return (&as_ymd) [1]; }
sub day   { return (&as_ymd) [2]; }

sub day_of_week {
    return ((${$_[0]} + 4) % 7);
}

#------------------------------------------------------------------------------
# the following methods are called by the overloaded operators, so they should
# not normally be called directly.
#------------------------------------------------------------------------------

sub _add {
    my ($date, $diff) = @_;

    if ($diff !~ /^-?\d+$/) {
	Carp::croak ("Date interval must be an integer");
    }
    my $new_date = bless (\ ($$date + $diff), ref($date));
    $new_date->default_format($date->default_format);
    return $new_date;
}

sub _subtract {
    my ($left, $right, $reverse) = @_;

    my $new_date;

    if ($reverse) {
	Carp::croak ("Can't subtract a date from a non-date");
    }
    if (ref($right) eq '' && $right =~ /^-?\d+$/) {
    $new_date = bless (\ ($$left - $right), ref($left));
    $new_date->default_format($left->default_format);
    return $new_date;
    }
    return ($$left - $$right);
}

sub _compare {
    my ($left, $right, $reverse) = @_;

    $right = $left->new($right) || _inval ($left, $right);
    return ($reverse ? $$right <=> $$left : $$left <=> $$right);
}

sub _eq {
    my ($left, $right) = @_;
    return (($right = $left->_new($right)) && $$right == $$left);
}

sub _ne {
    return (!&_eq);
}

1;

=head1 NAME

Date::Simple::NoXS - Pure Perl support for Date::Simple.

=head1 SYNOPSIS

    use Date::Simple;

=head1 DESCRIPTION

Used internally by Date::Simple.

=head1 SEE ALSO

L<Date::Simple>.

=cut


lib/Date/Simple.pm
# Date::Simple - a simple date object

package Date::Simple;

BEGIN {
    $VERSION = '3.03_03';
}

use Exporter ();
@ISA = ('Exporter');

@EXPORT_OK = qw(
  today ymd d8 leap_year days_in_month
  date date_fmt date_d8 date_iso
);

%EXPORT_TAGS = ( 'all' => \@EXPORT_OK );

# Try to load the C code.  If that fails, fall back to Date::Simple::NoXS.
if ( !defined(&_add) ) {
    my $err = $Date::Simple::NoXS;
    unless ($err) {

        # Use DynaLoader instead of XSLoader for pre-5.005.
        local ($@);
        local @ISA = ('DynaLoader');
        require DynaLoader;
        eval { __PACKAGE__->bootstrap($VERSION); };
        $err = $@;
    }
    if ($err) {
        $Date::Simple::NoXs = 1;
        require Date::Simple::NoXS;
    }
}

use strict;
use Carp ();
use overload
  '+'    => '_add',
  '-'    => '_subtract',
  '=='   => '_eq',
  '!='   => '_ne',
  '<=>'  => '_compare',
  'eq'   => '_eq',
  'ne'   => '_ne',
  'cmp'  => '_compare',
  'bool' => sub { 1 },
  '""'   => 'as_iso';

use Scalar::Util qw(refaddr reftype);
use warnings::register;
require Date::Simple::Fmt;
require Date::Simple::ISO;
require Date::Simple::D8;

sub d8 {

    # called as function
    if ( $#_ == 0 ) {
        return __PACKAGE__->_d8(@_);
    }

    # called as method
    else {
        if ( ref $_[0] eq 'SCALAR' ) {
            return $_[0]->SUPER::_d8(@_);
        }
        else {
            return $_[0]->_d8(@_);
        }
    }
}

sub today {
    if ( $#_ == -1 ) {
        return __PACKAGE__->_today(@_);
    }
    else {
        return shift->_today(@_);
    }
}

sub ymd {

    # called as function
    if ( $#_ == 2 ) {
        return __PACKAGE__->_ymd(@_);
    }

    # called as method
    else {
        if ( ref $_[0] eq 'SCALAR' ) {
            return $_[0]->SUPER::_ymd(@_);
        }
        else {
            return $_[0]->_ymd(@_);
        }
    }
}

sub _today {
    my ( $y, $m, $d ) = (localtime)[ 5, 4, 3 ];
    $y += 1900;
    $m += 1;
    return $_[0]->_ymd( $y, $m, $d );
}

sub _inval {
    my ($first);
    $first = shift;
    Carp::croak( "Invalid "
          . ( ref($first) || $first )
          . " constructor args: ('"
          . join( "', '", @_ )
          . "')" );
}

sub _new {
    my ( $that, @ymd ) = @_;

    my $class = ref($that) || $that;

    if ( @ymd == 1 ) {
        my $x = $ymd[0];
        if ( ref $x and reftype($x) eq 'ARRAY' ) {
            @ymd = @$x;
        }
        elsif ( UNIVERSAL::isa( $x, __PACKAGE__ ) ) {
            return ($x);
        }
        elsif ($x =~ /^(\d\d\d\d)-(\d\d)-(\d\d)$/
            || $x =~ /^(\d\d\d\d)(\d\d)(\d\d)$/ ) {
            @ymd = ( $1, $2, $3 );
        }
        else {
            return (undef);
        }
    }    # we fall through here...

    # note we can end up here is they pass in [] as the date
    return $class->_today() unless @ymd;

    # to get here, we had one arg which was split,
    # or 3 in the first place
    if ( @ymd == 3 ) {
        my $days = ymd_to_days(@ymd);
        return undef if !defined($days);
        return ( bless( \$days, $class ) );
    }

    $class->_inval(@ymd);
}

sub date { scalar __PACKAGE__->_new(@_) }

sub date_fmt {
    my $format = shift;
    my $obj    = Date::Simple::Fmt->_new(@_);
    $obj->default_format($format)
      if $obj;
    $obj;
}

sub date_d8  { scalar Date::Simple::D8->_new(@_) }
sub date_iso { scalar Date::Simple::ISO->_new(@_) }

# Same as date() but it's a method and croaks on error if called with
# one arg.
sub new {
    my ( $class, $date );

    $date = &_new;
    if ( !$date && scalar(@_) == 1 ) {
        Carp::croak( "'" . shift() . "' is not a valid ISO formated date" );
    }
    return ($date);
}

sub next { return ( $_[0] + 1 ); }
sub prev { return ( $_[0] - 1 ); }

sub _gmtime {
    my ( $y, $m, $d ) = days_to_ymd( ${ $_[0] } );
    $y -= 1900;
    $m -= 1;
    return ( 0, 0, 0, $d, $m, $y );
}

BEGIN {
    our $Standard_Format = "%Y-%m-%d";
    my %fmts = (    # Inside out parameter
        'Date::Simple'      => $Standard_Format,
        'Date::Simple::ISO' => $Standard_Format,
        'Date::Simple::D8'  => "%Y%m%d",
        'Date::Simple::Fmt' => $Standard_Format,
    );

    sub format {
        my ( $self, $format ) = @_;

        $format =
             $fmts{ refaddr($self) || '' }
          || $fmts{ ref($self) }
          || $Standard_Format
          if @_ == 1;

        return "$self" unless defined($format);
        require POSIX;
        local $ENV{TZ} = 'UTC+0';
        return POSIX::strftime( $format, _gmtime($self) );
    }

    sub strftime { &format }
    sub as_str   { &format }

    sub default_format {
        my ( $self, $val ) = @_;

        my $o = refaddr($self) || $self;

        if ( @_ > 1 ) {
            $fmts{$o} = $val;
            warnings::warnif "Setting class specific date format '$o' to" . "'"
              . ( defined $val ? $val : 'undef' ) . "'"
              unless ref $self;
        }

        return $fmts{$o} || $Standard_Format;
    }

    sub DESTROY {
        delete $fmts{ refaddr $_[0] };
    }
}

1;

=head1 NAME

Date::Simple - a simple date object

=head1 SYNOPSIS

    use Date::Simple ('date', 'today');

    # Difference in days between two dates:
    $diff = date('2001-08-27') - date('1977-10-05');

    # Offset $n days from now:
    $date = today() + $n;
    print "$date\n";  # uses ISO 8601 format (YYYY-MM-DD)

    use Date::Simple ();
    my $date  = Date::Simple->new('1972-01-17');
    my $year  = $date->year;
    my $month = $date->month;
    my $day   = $date->day;

    use Date::Simple (':all');
    my $date2 = ymd($year, $month, $day);
    my $date3 = d8('19871218');
    my $today = today();
    my $tomorrow = $today + 1;
    if ($tomorrow->year != $today->year) {
        print "Today is New Year's Eve!\n";
    }

    if ($today > $tomorrow) {
        die "warp in space-time continuum";
    }

    print "Today is ";
    print(('Sun','Mon','Tues','Wednes','Thurs','Fri','Satur')
          [$today->day_of_week]);
    print "day.\n";

    # you can also do this:
    ($date cmp "2001-07-01")
    # and this
    ($date <=> [2001, 7, 1])

=begin text

INSTALLATION

 If your system has the "make" program or a clone:

     perl Makefile.PL
     make
     make test
     make install

 If you lack "make", copy the "lib/Date" directory to your module
 directory (run "perl -V:sitelib" to find it).

 If "make test" fails, perhaps it means your system can't compile C
 code.  Try:

     make distclean
     perl Makefile.PL noxs
     make
     make test
     make install

 This will use the pure-Perl implementation.

=end text

=head1 DESCRIPTION

Dates are complex enough without times and timezones.  This module may
be used to create simple date objects.  It handles:

=over 4

=item Validation.

Reject 1999-02-29 but accept 2000-02-29.

=item Interval arithmetic.

How many days were between two given dates?  What date comes N days
after today?

=item Day-of-week calculation.

What day of the week is a given date?

=item Transparent date formatting.

How should a date object be formatted.

=back

It does B<not> deal with hours, minutes, seconds, and time zones.

A date is uniquely identified by year, month, and day integers within
valid ranges.  This module will not allow the creation of objects for
invalid dates.  Attempting to create an invalid date will return
undef.  Month numbering starts at 1 for January, unlike in C and Java.
Years are 4-digit.

Gregorian dates up to year 9999 are handled correctly, but we rely on
Perl's builtin C<localtime> function when the current date is
requested.  On some platforms, C<localtime> may be vulnerable to
rollovers such as the Unix C<time_t> wraparound of 18 January 2038.

Overloading is used so you can compare or subtract two dates using
standard numeric operators such as C<==>, and the sum of a date object
and an integer is another date object.

Date::Simple objects are immutable.  After assigning C<$date1> to
C<$date2>, no change to C<$date1> can affect C<$date2>.  This means,
for example, that there is nothing like a C<set_year> operation, and
C<$date++> assigns a new object to C<$date>.

This module contains various undocumented functions.  They may not be
available on all platforms and are likely to change or disappear in
future releases.  Please let the author know if you think any of them
should be public.

=head2 Controlling output format.

As of version 3.0 new ways of controlling the output formats of Date::Simple
objects has been provided. However Date::Simple has traditionally provided
few ways of stringification, a primary one via the format() method and another
primary one via direct stringification. However the later is currently
implemented as an XS routine and the former is implemented through a perl routine.
This means that using format() is more expensive than stringification and
that the stringification format is class specific.

In order to alleviate some of these problems a new mechanism has been introduced
to Date::Simple that allows for a per object level format default. In addition
a set of utility classes that have different stringification overloads provided.
These classes are simple subclasses of Date::Simple and beside the default format()
and the overloaded stringification behaviour are identical to Date::Simple. In fact
one is totally identical to Date::Simple and is provided mostly for completeness.

The classes included are:

=over 4

=item Date::Simple::ISO

Identical to Date::Simple in every respect but name.

=item Date::Simple::D8

Uses the D8 format (%Y%m%d) as the default format for printing. Uses XS for the
overloaded stringification.

=item Date::Simple::Fmt

Uses the perl implemented format() as the default stringification mechanism. The first
argument to the constructor is expected to be the format to use for the object.

=back

B<NOTE> its important to remember that the primary difference between the behaviour
of objects of the different classes is how they are stringified when quoted, and what
date format is used by default when the format() method is called. Nothing else differs.

=head1 CONSTRUCTORS

Several functions take a string or numeric representation and generate
a corresponding date object.  The most general is C<new>, whose
argument list may be empty (returning the current date), a string in
format YYYY-MM-DD or YYYYMMDD, a list or arrayref of year, month, and
day number, or an existing date object.

=over 4

=item Date::Simple->new ([ARG, ...])

=item date ([ARG, ...])

    my $date = Date::Simple->new('1972-01-17');

The C<new> method will return a date object if the values passed in
specify a valid date.  (See above.)  If an invalid date is passed, the
method returns undef.  If the argument is invalid in form as opposed
to numeric range, C<new> dies.

The C<date> function provides the same functionality but must be
imported or qualified as C<Date::Simple::date>.  (To import all public
functions, do C<use Date::Simple (':all');>.)  This function returns
undef on all invalid input, rather than dying in some cases like
C<new>.

=item date_fmt (FMT,[ARG, ...])

Equivelent to C<date> but creates a Date::Simple::Fmt object instead. The
format is expected to be a valid POSIX::strftime format string.

=item date_iso ([ARG, ...])

Identical to C<date> but creates a Date::Simple::ISO object instead.

=item date_d8 ([ARG, ...])

Equivelent to C<date> but creates a Date::Simple::D8 object instead.

=item today()

Returns the current date according to C<localtime>.

B<Caution:> To get tomorrow's date (or any fixed offset from today),
do not use C<today + 1>.  Perl parses this as C<today(+1)>.  You need
to put empty parentheses after the function: C<today() + 1>.

=item ymd (YEAR, MONTH, DAY)

Returns a date object with the given year, month, and day numbers.  If
the arguments do not specify a valid date, undef is returned.

Example:

    use Date::Simple ('ymd');
    $pbd = ymd(1987, 12, 18);

=item d8 (STRING)

Parses STRING as "YYYYMMDD" and returns the corresponding date object,
or undef if STRING has the wrong format or specifies an invalid date.

Example:

    use Date::Simple ('d8');
    $doi = d8('17760704');

Mnemonic: The string matches C</\d{8}/>.  Also, "d8" spells "date", if
8 is expanded phonetically.

=back

=head1 INSTANCE METHODS

=over 4

=item DATE->next

    my $tomorrow = $today->next;

Returns an object representing tomorrow.

=item DATE->prev

   my $yesterday = $today->prev;

Returns an object representing yesterday.

=item DATE->year

    my $year  = $date->year;

Return the year of DATE as an integer.

=item DATE->month

    my $month = $date->month;

Return the month of DATE as an integer from 1 to 12.

=item DATE->day

    my $day   = $date->day;

Return the DATE's day of the month as an integer from 1 to 31.

=item DATE->day_of_week

Return a number representing DATE's day of the week from 0 to 6, where
0 means Sunday.

=item DATE->as_ymd

    my ($year, $month, $day) = $date->as_ymd;

Returns a list of three numbers: year, month, and day.

=item DATE->as_d8

Returns the "d8" representation (see C<d8>), like
C<$date-E<gt>format("%Y%m%d")>.

=item DATE->as_iso

Returns the ISO 8601 representation of the date (eg '2004-01-01'),
like C<$date-E<gt>format("%Y-%m-%d")>. This is in fact the default
overloaded stringification mechanism and is provided mostly so
other subclasses with different overloading can still do fast
ISO style date output.

=item DATE->as_str ([STRING])

=item DATE->format ([STRING])

=item DATE->strftime ([STRING])

These functions are equivalent.  Return a string representing the
date, in the format specified.  If you don't pass a parameter, the default
date format for the object is used if one has been specified, otherwise
uses the default date format for the class the object is a member of, or as
a last fallback uses the $Date::Simple::Standard_Format which is changeable,
but probably shouldn't be modified. See C<default_format> for details.

    my $change_date = $date->format("%d %b %y");
    my $iso_date1 = $date->format("%Y-%m-%d");
    my $iso_date2 = $date->format;

The formatting parameter is similar to one you would pass to
strftime(3).  This is because we actually do pass it to strftime to
format the date.  This may result in differing behavior across
platforms and locales and may not even work everywhere.

=item DATE->default_format ([FORMAT])

This method sets or gets the default_format for the DATE object or class
that it is called on.

=back

=head1 OPERATORS

Some operators can be used with Date::Simple instances.  If one side
of an expression is a date object, and the operator expects two date
objects, the other side is interpreted as C<date(ARG)>, so an array
reference or ISO 8601 string will work.

=over 4

=item DATE + NUMBER

=item DATE - NUMBER

You can construct a new date offset by a number of days using the C<+>
and C<-> operators.

=item DATE1 - DATE2

You can subtract two dates to find the number of days between them.

=item DATE1 == DATE2

=item DATE1 < DATE2

=item DATE1 <=> DATE2

=item DATE1 cmp DATE2

=item etc.

You can compare two dates using the arithmetic or string comparison
operators.  Equality tests (C<==> and C<eq>) return false when one of
the expressions can not be converted to a date.  Other comparison
tests die in such cases.  This is intentional, because in a sense, all
non-dates are not "equal" to all dates, but in no sense are they
"greater" or "less" than dates.

=item DATE += NUMBER

=item DATE -= NUMBER

You can increment or decrement a date by a number of days using the +=
and -= operators.  This actually generates a new date object and is
equivalent to C<$date = $date + $number>.

=item "$date"

You can interpolate a date instance directly into a string, in the
format specified by ISO 8601 (eg: 2000-01-17) for Date::Simple and
Date::Simple::ISO, for Date::Simple::D8 this is the same as calling
as_d8() on the object, and for Date::Simple::Fmt this is the same as
calling format() on the object.

=back

=head1 UTILITIES

=over 4

=item leap_year (YEAR)

Returns true if YEAR is a leap year.

=item days_in_month (YEAR, MONTH)

Returns the number of days in MONTH, YEAR.

=back

=over 4

=item leap_year (YEAR)

Returns true if YEAR is a leap year.

=item days_in_month (YEAR, MONTH)

Returns the number of days in MONTH, YEAR.

=back


=head1 AUTHOR

    Marty Pauley <marty@kasei.com>
    John Tobey <jtobey@john-edwin-tobey.org>
    Yves Orton <demerphq@hotmail.com>

=head1 COPYRIGHT

      Copyright (C) 2001  Kasei.
      Copyright (C) 2001,2002 John Tobey.
      Copyright (C) 2004 Yves Orton.

      This program is free software; you can redistribute it and/or
      modify it under the terms of either:

      a) the GNU General Public License;
         either version 2 of the License, or (at your option) any later
         version.  You should have received a copy of the GNU General
         Public License along with this program; see the file COPYING.
         If not, write to the Free Software Foundation, Inc., 59
         Temple Place, Suite 330, Boston, MA 02111-1307 USA

      b) the Perl Artistic License.

      This program is distributed in the hope that it will be useful,
      but WITHOUT ANY WARRANTY; without even the implied warranty of
      MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

=head1 SEE ALSO

L<Date::Simple::Fmt> L<Date::Simple::ISO> L<Date::Simple::D8>
and of course L<perl>

=cut
